base/strings/util.jl

```diff
@@ -513,6 +513,126 @@ function rpad(
     r == 0 ? stringfn(s, p^q) : stringfn(s, p^q, first(p, r))
 end
 
+"""
+    rtrunc(str::AbstractString, w::Int, replace_str::AbstractString = "…")
+
+Truncate `str` to at most `w` characters (see [`textwidth`](@ref)), replacing the last characters
+with `replace_str` if necessary. The default replacement string is "…".
+
+# Examples
+```jldoctest
+julia> s = rtrunc("🍕🍕 I love 🍕", 10)
+"🍕🍕 I lo…"
+
+julia> textwidth(s)
+10
+
+julia> rtrunc("foo", 3)
+"foo"
+```
+
+!!! compat "Julia 1.12"
+    This function was added in Julia 1.12.
+
+See also [`ltrunc`](@ref) and [`ctrunc`](@ref).
+"""
+function rtrunc(str::AbstractString, w::Int, replace_str::AbstractString = "…")
+    if textwidth(str) <= w
+        return str
+    else
+        i, accwidth = firstindex(str), textwidth(replace_str)
+        while true
+            tw = textwidth(str[i])
+            accwidth + tw <= w || break # no need to check if we go out of bounds because of the first if branch
+            accwidth += tw
+            i = nextind(str, i)
+        end
+        return str[firstindex(str):prevind(str, i)] * replace_str
```

Comment:
```suggestion
        @views return str[firstindex(str):prevind(str, i)] * replace_str
```
(avoids an allocation)